### PR TITLE
port to python3

### PR DESCRIPTION
**Pull request by kobbled**
With python finally dropping support for python2 next year it might be a good idea to patch for python3. Only potential issue is that newer version of python use the py launcher executable instead of calling python with 'python' or 'python3' so the .cmd file also needed to be updated, which will not work if python launcher is not installed.

### CODE REVIEW

File: ktransw.py [hunk 1]
```diff
@@ -243,7 +243,7 @@ def main():
                         hdr_path = os.path.join(hdr_dir, hdr_path)
                         logger.debug("Found {0} in '{1}'".format(hdr, hdr_dir))
 
-                    except ValueError as e:
+                    except (ValueError) as e:
```
Comment (gavanderhoorn) [inline ktransw.py]: A quick Google doesn't explain to me why the parentheses should be added here. Is this a Python3 best practice?

File: ktransw.py [hunk 2]
```diff
@@ -312,7 +312,9 @@ def main():
             # TODO: we loose stdout/stderr interleaving here
             # TODO: the error messages refer to lines in the temporary,
             # preprocessed KAREL source file, not the original one.
-            sys.stdout.write(pstdout.replace(dname, os.path.dirname(kl_file)) + '\n')
+            pstdout = pstdout.decode('utf-8')
+            ktrans_out = pstdout.replace(dname, os.path.dirname(kl_file)) + '\n'
```
Comment (gavanderhoorn) [inline ktransw.py]:
```suggestion
            ktrans_out = pstdout.decode('utf-8').replace(dname, os.path.dirname(kl_file)) + '\n'
```

Could be nice to chain this (and then remove the `pstdout = pstdout.decode('utf-8')` line above).

File: ktransw.cmd [hunk 3]
```diff
@@ -1,17 +1,33 @@
...
+
 REM distributed under the License is distributed on an "AS IS" BASIS,
+
 REM WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+
 REM See the License for the specific language governing permissions and
+
 REM limitations under the License.
+
 REM
```
Comment (gavanderhoorn) [inline ktransw.cmd]: This seems to introduce a lot of empty lines. Is this a line-ending conflict?

**Review by gavanderhoorn — COMMENTED** [review]
Thanks for the PR. A few in-line comments and a more general one.

Is the Python launcher really needed? I've not used it myself, but from what I gather from reading some sites it seems to be an optional part of Python 3 (it seems to not even always be included in an install). Calling `python.exe` should still work, it would depend on the `PATH` of the user which Python is then used.

I'm not against using the launcher, but would like to understand the situation a bit better.

File: ktransw.py (hunk 1 above)
Comment (kobbled) [reply ktransw.py]: I believe it was throwing an error without it. I had to google that myself. Should double check though.

File: ktransw.py (hunk 2 above)
Comment (kobbled) [reply ktransw.py]: definitely. I think I was just testing and forgot to get rid of the extra line.

File: ktransw.cmd (hunk 3 above)
Comment (kobbled) [reply ktransw.cmd]: Rossum.cmd was fine in my editor. ktransw.cmd had no carriage returns for me, just one line.

**Comment by kobbled** [comment]
Haven't tested using just 'python'. I still have a python2 and python3 on my machine so using just python will always give me python2. The py launcher does seem like a good idea to separate the two though. Don't know what the future entails for it. Need to investigate this further.

File: ktransw.cmd (hunk 3 above)
Comment (gavanderhoorn) [reply ktransw.cmd]: >  ktransw.cmd had no carriage returns for me, just one line

You may have a global setting for line-endings that is (trying to) overriding what the repository is configured for. See whether [Dealing with line endings](https://help.github.com/en/articles/dealing-with-line-endings) sheds some light on this.

**Comment by gavanderhoorn** [comment]
> Haven't tested using just 'python'. I still have a python2 and python3 on my machine so using just python will always give me python2.

So is `python2` in a directory that is before the `python3` directory in your `PATH`? That would/could explain this.

> The py launcher does seem like a good idea to separate the two though. Don't know what the future entails for it.

Python 2 doesn't have the launcher, so even if both intepreters are on the `PATH` it would only ever be found in the Python 3 location.

As I wrote above: I'm not against standardising on it, just want to understand it.

**Comment by kobbled** [comment]
Thinking about it though I don't think the py launcher is standardizing. My python setup on windows is a bit of a mess for me so the py launcher helps out, but that is my fault for not using environments when needed. Embedding the python launcher into a batch file calling ktransw would be very confusing, and cause issues. I have reverted the changes to the cmd file.

**Comment by kobbled** [comment]
I have updated the changes you suggested above to this branch!

**Comment by gavanderhoorn** [comment]
Hi @kobbled. You seem to have pushed some additional commits to this branch.

That's OK, but I'm not sure they should be part of this PR.

I noticed you've been really making serious use of `gpp` and its capabilities. Could you enable *issues* on your fork of `ktransw_py`? I'd like to post a comment about your implementation of objects and classes.

**Comment by kobbled** [comment]
@gavanderhoorn, I opened up the issues on the fork. Yes, it been a while since I've come back to this so I forgot this PR was still open and that it wasn't on a separate branch. If you want to close this, I can create a branch up to a reasonable commit and resubmit. With extra covid time I've been experimenting with gpp, and your libsimplemsg got me thinking it wouldn't be to difficult to implement crude objects into ktransw with some hand waving with gpp. I'm just getting around to understanding how to use a preprocessor in more depth than the typical uses you see in c.  There are definitely some not good practices I used to achieve this, but I just wanted to see if I could get it to work with the experience that I have.

**Pull request closed without merging**